theseus/core/vectorizer.py

```diff
@@ -0,0 +1,291 @@
+from collections import defaultdict
+from typing import Dict, List, Optional, Set, Tuple
+
+import torch
+
+from .cost_function import CostFunction, _register_vars_in_list
+from .objective import Objective
+from .variable import Variable
+
+_CostFunctionSchema = Tuple[str, ...]
+
+
+def _get_cost_function_schema(cost_function: CostFunction) -> _CostFunctionSchema:
+    def _fullname(obj) -> str:
+        return f"{obj.__module__}.{obj.__class__.__name__}"
+
+    def _varinfo(var) -> str:
+        return f"{_fullname(var)}{tuple(var.shape[1:])}"
+
+    return (
+        (_fullname(cost_function),)
+        + tuple(_varinfo(v) for v in cost_function.optim_vars)
+        + tuple(_varinfo(v) for v in cost_function.aux_vars)
+        + (_fullname(cost_function.weight),)
+        + tuple(_varinfo(v) for v in cost_function.weight.aux_vars)
+    )
+
+
+# This wrapper allows the weighted error and jacobians to be replaced by pre-computed
+# values. The vectorization class will compute these in a batch for all cost function,
+# then populate wrappers for each so they can be served to the linearization classes.
+class _CostFunctionWrapper(CostFunction):
+    def __init__(self, cost_fn: CostFunction, name: Optional[str] = None):
+        if name is None:
+            name = f"wrapper({cost_fn.name})"
+        super().__init__(cost_fn.weight, name=name)
+        self.cost_fn = cost_fn
+        _register_vars_in_list(self, cost_fn.optim_vars, is_optim=True)
+        _register_vars_in_list(self, cost_fn.aux_vars, is_optim=False)
+        self._cached_error: Optional[torch.Tensor] = None
+        self._cached_jacobians: Optional[List[torch.Tensor]] = None
+
+    def error(self) -> torch.Tensor:
+        return self.cost_fn.error()
+
+    def jacobians(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        return self.cost_fn.jacobians()
+
+    def weighted_error(self) -> torch.Tensor:
+        if self._cached_error is None:
+            return super().weighted_error()
+        return self._cached_error
+
+    def weighted_jacobians_error(
+        self,
+    ) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        if self._cached_jacobians is None:
+            return super().weighted_jacobians_error()
+        assert self._cached_error is not None
+        return self._cached_jacobians, self._cached_error
+
+    def dim(self) -> int:
+        return self.cost_fn.dim()
+
+    def _copy_impl(self, new_name: Optional[str] = None) -> "_CostFunctionWrapper":
+        return _CostFunctionWrapper(self.cost_fn, name=new_name)
+
+    def __str__(self) -> str:
+        return f"Wrapper({str(_get_cost_function_schema(self.cost_fn))})"
+
+    def __repr__(self) -> str:
+        return str(self)
```

Comment:
Is it guaranteed that `self._cached_error=None` and `self._cached_jacobians=None` whenever either `optim_vars` or `aux_vars` are updated?

Reply:
That's a great observation. In fact, this is actually only guaranteed if `Objective.update()` is used, but not if variables are updated via their stand alone `update` methods. This is why [this](https://github.com/facebookresearch/theseus/pull/205/files#diff-73a97268e17aeaa7f1e018f6f77e917666909b415d6d9567bea88977e53a576bR397) line is needed after the inner loop var update. 

Maybe the safest way to prevent unwanted behavior is (by default) enable the vectorized caches only when optimization is called via `TheseusLayer`?

Reply:
I agree. As long as users have to no access to `CostFunctionWrapper` and developers are aware of this, it should be safe.

Reply:
> enable the vectorized caches only when optimization is called via `TheseusLayer`
Is this already incorporated?

Reply:
This is already incorporated via the private iterator. That's the only way to use the vectorization results.